Source/Core/DiscIO/VolumeVerifier.cpp

```diff
@@ -39,6 +49,276 @@
 
 namespace DiscIO
 {
+std::once_flag RedumpVerifier::m_gc_download_flag;
+std::once_flag RedumpVerifier::m_wii_download_flag;
+RedumpVerifier::DownloadStatus RedumpVerifier::m_gc_download_status;
+RedumpVerifier::DownloadStatus RedumpVerifier::m_wii_download_status;
+
+void RedumpVerifier::Start(const Volume& volume)
+{
+  // We use GetGameTDBID instead of GetGameID so that Datel discs will be represented by an empty
+  // string, which matches Redump not having any serials for Datel discs.
+  m_game_id = volume.GetGameTDBID();
+  if (m_game_id.size() > 4)
+    m_game_id = m_game_id.substr(0, 4);
+
+  m_revision = volume.GetRevision().value_or(0);
+  m_disc_number = volume.GetDiscNumber().value_or(0);
+  m_size = volume.GetSize();
+
+  const DiscIO::Platform volume_type = volume.GetVolumeType();
+
+  m_future = std::async(std::launch::async, [this, volume_type]() -> std::vector<PotentialMatch> {
+    std::string platform;
+    std::once_flag* download_flag;
+    DownloadStatus* download_status;
+    switch (volume_type)
+    {
+    case Platform::GameCubeDisc:
+      platform = "gc";
+      download_flag = &m_gc_download_flag;
+      download_status = &m_gc_download_status;
+      break;
+
+    case Platform::WiiDisc:
+      platform = "wii";
+      download_flag = &m_wii_download_flag;
+      download_status = &m_wii_download_status;
+      break;
+
+    default:
+      m_result.status = Status::Error;
+      return {};
+    }
+
+    std::call_once(*download_flag, DownloadDatfile, platform, download_status);
+
+    switch (*download_status)
+    {
+    case DownloadStatus::Success:
+      return ScanDatfile(ReadDatfile(platform));
+
+    case DownloadStatus::SystemNotPublic:
+      m_result = {Status::Error, Common::GetStringT("Wii data is not public yet")};
+      return {};
+
+    case DownloadStatus::Fail:
+    default:
+      m_result = {Status::Error, Common::GetStringT("Failed to connect to Redump.org")};
+      return {};
+    }
+  });
+}
+
+static std::string GetPathForPlatform(const std::string& platform)
+{
+  return File::GetUserPath(D_REDUMPCACHE_IDX) + DIR_SEP + platform + ".zip";
+}
+
+void RedumpVerifier::DownloadDatfile(const std::string& platform, DownloadStatus* status_out)
+{
+  Common::HttpRequest request;
+
+  const std::optional<std::vector<u8>> result =
+      request.Get("http://redump.org/datfile/" + platform + "/serial,version",
+                  {{"User-Agent", Common::scm_rev_str}});
+
+  if (!result)
+  {
+    *status_out = DownloadStatus::Fail;
+    return;
+  }
+
+  if (result->size() > 1 && (*result)[0] == '<' && (*result)[1] == '!')
```

Comment:
While redump does present a `DOCTYPE` declaration (for XHTML 1.1), I wonder if checking for the open angle bracket would be enough (so it still works when they do). But then again, nothing mandates that it really needs to be starting with that. It could also have blanks, a BOM or other stuff that doesn't affect the DOM (altho, the majority _would_ indeed start with a `<` there when it's not whitespace).

Should we perhaps be checking the `Content-type` header instead (assuming we can somehow get it from the response without jumping too many hoops); or check for the PK3 Zip header instead (assuming everything else is _probably_ plain-text-ish)?

...the more I think about this (and type into the comment box), the more I think my comment is useless and won't being any change to this section of code...mainly because all the alternatives are just as bad.

Reply:
> While redump does present a `DOCTYPE` declaration (for XHTML 1.1), I wonder if checking for the open angle bracket would be enough (so it still works when they do).

So it still works when they do what?

Reply:
Remove the `DOCTYPE` (altho, the only thing they should be switching to at this point is potentially HTML5, which also has one), add copyright comment or whatever (which _also_ starts with `<!`) or...dunno, some other CDN shenanigans that leave the content same-ish but minutely mess with the first two bytes.

I suppose we should just drop the what-if and deal with it if/when it happens.

Reply:
Yeah. I think the chance of them changing how the error message is written is larger than them making the page not start with `<!`, really. But redump doesn't tend to get a lot of changes anyway...

Hopefully they'll make the Wii datfile public soon, and then we can just remove this piece of code.